Fix #2297 - Added Speech Recognition to TextInput Field

bansalnitish opened the pull request:
Fixes #2297 
- Speech recognition is added for single line and multiple line text input.
- This speech Recognition works only for browsers that support webkitSpeechRecognition i.e Chrome only. So the option for choosing language is disabled for all others browsers and is enabled only for chrome. 
- UI is completed.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes.
- [x] The PR explanation includes the words "Fixes #bugnum: ...".
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

bansalnitish committed:
Added Speech Recognition to TextInput Field

bansalnitish commented:
@tjiang11 can you just a give a pass on this PR? Build is failing because I have changed the input format and test for that needs to be updated in `schema_utils.py` file I think. I have checked the code multiple times it works fine according to me. I have also tried to solve the build issues but I need help here.
Thanks

tjiang11 commented on extensions/interactions/TextInput/directives/TextInput.js:
```diff
@@ -42,19 +42,28 @@ oppia.directive('oppiaInteractiveTextInput', [
+            $scope.schema.ui_config.speechRecognitionLanguage = '';
+          } else {
+            $scope.schema.ui_config.speechRecognitionLanguage =
+            $scope.speechrelang;
```
Indent this line by two spaces.

tjiang11 commented on extensions/interactions/TextInput/directives/TextInput.js:
```diff
@@ -42,19 +42,28 @@ oppia.directive('oppiaInteractiveTextInput', [
             $attrs.placeholderWithValue);
           $scope.rows = (
             HtmlEscaperService.escapedJsonToObj($attrs.rowsWithValue));
+          $scope.speechrelang = HtmlEscaperService.escapedJsonToObj(
```
$scope.speechRecognitionLanguage, or something more understandable

tjiang11 commented on core/templates/dev/head/components/forms/schema_editors/schema_based_unicode_editor_directive.html:
```diff
@@ -1,4 +1,5 @@
 <span>
+ <span ng-if="!getCodingMode() && getRows()">
   <textarea ng-if="!getCodingMode() && getRows()"
```
Is this `ng-if` redundant?

bansalnitish replied:
Yes it was !

tjiang11 commented on core/templates/dev/head/components/forms/schema_editors/schema_based_unicode_editor_directive.html:
```diff
@@ -8,6 +9,7 @@
             ng-blur="onInputBlur()"
             ng-focus="onInputFocus()">
   </textarea>
+  </span>
```
De-indent by two spaces.

tjiang11 replied:
@bansalnitish please reply to comments to indicate you have addressed them

bansalnitish replied:
Sorry for the issue @tjiang11 :sweat_smile: I will fix this!

bansalnitish replied:
I have fixed the indentation!

tjiang11 commented on core/templates/dev/head/components/forms/schema_editors/schema_based_unicode_editor_directive.html:
```diff
@@ -17,6 +19,15 @@
 
+   <span ng-if="allowSpeechRecognition && !getCodingMode()" ng-click="toggleSpeechRecognition()">
+      <img ng-if="!isCurrentlyRecording"
+           src="https://raw.githubusercontent.com/FezVrasta/material-design-icons-svg/master/av/3x_ios/ic_mic_black_24dp.png"
```
We should put the image into assets and then load it from there rather than from some external link. Also, I think font awesome might have suitable images?

tjiang11 commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -173,6 +173,18 @@ body {
   width: 100%;
 }
 
+.oppia-editable-section .oppia-unicode-form-mic-button {
```
Alphabetize the attributes.

tjiang11 commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -173,6 +173,18 @@ body {
+  height: 40px;
+}
+
+.conversation-skin-tutor-card .oppia-unicode-form-mic-button {
```
Ditto

tjiang11 commented on core/templates/dev/head/components/forms/schema_editors/SchemaBasedChoicesEditorDirective.js:
```diff
@@ -37,7 +37,11 @@ oppia.directive('schemaBasedChoicesEditor', [
       restrict: 'E',
       compile: NestedDirectivesRecursionTimeoutPreventionService.compile,
-      controller: ['$scope', function($scope) {
+      controller: ['$scope', '$window', function($scope, $window) {
```
Why do we have to modify SchemaBasedChoicesEditor? Sorry I'm not too familiar with this.

bansalnitish replied:
@tjiang11 I wanted to disable the choices for language recognition in browsers other than chrome. So, I modified the ``isDisabled()`` when the schema was used for speech recognition and the browser did not support speech recognition. I tested this locally and it worked fine!

tjiang11 replied:
Sorry, I'm a little confused what these choices are referring to. Is there any way you can take a screenshot of this? Is it possible for me to change the language I want to the speech recognition to be in from the learner view?

bansalnitish replied:
There are choices for language in speech recognition. I wanted them to be enabled in chrome since only chrome supports ``webkitSpeechRecognition`` and other browsers don't.
Here are the screenshots for the same:
![screenshot from 2018-02-28 00-08-51](https://user-images.githubusercontent.com/22434689/36747779-ec344e9c-1c1b-11e8-9652-9a4c4c3cfb8c.png)
![screenshot from 2018-02-28 00-09-37](https://user-images.githubusercontent.com/22434689/36747789-f245d8fa-1c1b-11e8-82f7-4c9f97783091.png)
The first image is in a Chrome browser and the second one is in Firefox.

tjiang11 replied:
Okay, so in the Firefox screenshot, it looks as if the language is set to English and unalterable. But this isn't actually the case, since speech recognition is completely disabled, in which case we just shouldn't even show this field. Also though, this doesn't really make sense from a creator perspective. I could be creating an exploration in Firefox but expect learners using Chrome to play my exploration, so we should not be disabling and enabling the feature based on what browser the creator is using.

bansalnitish replied:
Yes, that is right @tjiang11. I will remove the disable feature!

tjiang11 commented on core/templates/dev/head/components/forms/schema_editors/SchemaBasedChoicesEditorDirective.js:
```diff
@@ -37,7 +37,11 @@ oppia.directive('schemaBasedChoicesEditor', [
       compile: NestedDirectivesRecursionTimeoutPreventionService.compile,
-      controller: ['$scope', function($scope) {
+      controller: ['$scope', '$window', function($scope, $window) {
+        $scope.isDisabled = function() {
```
Does this override the `isDisabled()` that is passed in on scope?

bansalnitish replied:
Yes, I think it was. I have fixed that!

tjiang11 commented on core/templates/dev/head/components/forms/schema_editors/SchemaBasedUnicodeEditorDirective.js:
```diff
@@ -121,6 +123,57 @@ oppia.directive('schemaBasedUnicodeEditor', [
+
+          $scope.isCurrentlyRecording = false;
+
+          var recognition;
```
Let's set this to `null`.

tjiang11 reviewed (changes requested):
Thanks @bansalnitish ! Pretty cool stuff :)

UI: Can you put the microphone icon/image inside the text input bar? Kind of like this: 
![image](https://user-images.githubusercontent.com/11807091/36695247-c4da2afa-1b0e-11e8-871a-9f185ff9e638.png)
That way, it's more clear that it's associated with the input.

Also, I think the icon is a little unclear--it's not entirely obvious that an open mic means recording and a filled in mic means not recording. I think maybe it would be clearer if the mic color changed to red. It could also be useful to add some sort of prompt such as "Speak now" or similar.

bansalnitish commented:
@tjiang11 I have made the changes. All thanks to your suggestions, the UI looks pretty better now!

tjiang11 commented on core/templates/dev/head/components/forms/schema_editors/SchemaBasedChoicesEditorDirective.js:
```diff
@@ -37,7 +37,11 @@ oppia.directive('schemaBasedChoicesEditor', [
       compile: NestedDirectivesRecursionTimeoutPreventionService.compile,
-      controller: ['$scope', function($scope) {
+      controller: ['$scope', '$window', function($scope, $window) {
+        if ($scope.schema().value && $scope.schema().value === 'srl' &&
```
In what case is it not equal to 'srl'?

bansalnitish replied:
I have defined value in the schema for Speech recognition to be equal to ``srl``. So, it is only equal to ``srl`` if the schema is for Speech recognition language.
https://github.com/oppia/oppia/pull/4740/files#diff-c24716e5861ee68472a4ce82b413a6a9R68
Using this value check I was disabling this schema in browsers other than Chrome.

tjiang11 commented on core/templates/dev/head/components/forms/schema_editors/SchemaBasedChoicesEditorDirective.js:
```diff
@@ -37,7 +37,11 @@ oppia.directive('schemaBasedChoicesEditor', [
+      controller: ['$scope', '$window', function($scope, $window) {
+        if ($scope.schema().value && $scope.schema().value === 'srl' &&
+            !$window.webkitSpeechRecognition) {
+          $scope.isDisabled = true;
```
isDisabled should be a function as it's called using `isDisabled()` from the template

bansalnitish replied:
Ah! my bad. I have fixed that.

tjiang11 reviewed (changes requested):
Design: When it's recording, I think the microphone itself should be red, rather than have a red background. Also, there should not be a slash through the microphone, as this usually indicated the microphone is muted, and we don't want to give that impression as it could be really confusing. Also, could you make the microphone icon just a bit smaller so that it's not right up against the edges on the input box? Thanks!

bansalnitish commented:
@tjiang11 I have updated the UI as per your request.

tjiang11 commented:
Hi Nithesh, 

- I'm getting this in the creator view. I think it should go inside the text input to match the learner view.
  ![image](https://user-images.githubusercontent.com/11807091/36826996-d1b39e2e-1cde-11e8-84fb-83ccdd4e3b56.png)
- The icon style/shape shouldn't really change when recording in the learner view, other than the color.
- Thank you for clearing up what choices it was referring to. Two things about this: 1) Users can't understand language codes. It should be 'English' instead of 'en'. 2) Does it make sense for the creator to be manually defining this for each interaction? Wouldn't it make more sense to just have it match the exploration's language?

bansalnitish commented:
Hi @tjiang11, I agree with you that it makes more sense to match the speech recognition language to the exploration language. Should I remove the choices feature or add the default as the exploration language instead of English?
Also, I have removed the disable feature and fixed the icons.

tjiang11 commented:
I think it would be best to simply remove the option.

bansalnitish commented:
@tjiang11 I have removed the choices for language and used the exploration language as the speech recognition language.

bansalnitish commented:
@tjiang11 Please have a look now !

tjiang11 commented:
Hi @nithusha21 , thanks!

It seems that the icon is messing up the animation between cards. To reproduce:

1. Login as admin and load the Welcome to Oppia! exploration
2. Go to /explore/0
3. Answer the first card.
4. Now, press 'Continue'. Notice that the card gets extremely large and then small again when the next card appears. The expected behavior is that the card gradually increases to the next card's size.

nithusha21 commented:
Hi @tjiang11 I think you tagged the wrong Nithesh :P The owner of this PR is @bansalnitish

tjiang11 commented:
Haha oops sorry

bansalnitish commented:
Sorry for the delay @tjiang11 !
I have updated the PR. It removes that card enlargement error.
Thank you for the review

tjiang11 commented on core/templates/dev/head/components/forms/schema_editors/SchemaBasedUnicodeEditorDirective.js:
```diff
@@ -121,6 +123,58 @@ oppia.directive('schemaBasedUnicodeEditor', [
+            recognition.interimResults = true;
+            var explorationLanguageCode = (
+                    ExplorationPlayerService.getExplorationLanguageCode());
+            recognition.lang = explorationLanguageCode;
```
What happens if speech recognition is not supported for the exploration's language code?

bansalnitish replied:
Should I add a default language English for such cases?

tjiang11 replied:
In that case, I would say the option to use speech recognition input simply should not be available

bansalnitish replied:
Thanks for the suggestion! I will make the changes!

bansalnitish replied:
Hi @tjiang11. I have checked for the language support in webkit speech recognition. It supports all the languages we currently have. Also, there is no specific way to obtain a list of supported languages. In case the language is not supported it automatically switches to English. So, shall I create a separate list of supported languages for checking this? Can you please provide some suggestions on this!
Here is a link for the same:https://stackoverflow.com/questions/14257598/what-are-language-codes-in-chromes-implementation-of-the-html5-speech-recogniti

tjiang11 replied:
Yes, I think it'd be best if we explicitly define the languages we support for speech recognition. I think it'd be sufficient to just do English for now, but feel free to do more. We don't want it defaulting because if we do end up having explorations in languages that aren't supported, such as perhaps a purely Hinglish exploration, then we don't want it to be taking in input in English.

tjiang11 replied:
Is it possible to detect whether a specific language is supported?

bansalnitish replied:
Hey @tjiang11 
I searched for this and I did not find an explicit way to check if a language is supported. 

Also, the languages we currently have as options in an exploration's language are all supported by webkit speech recognition. 

So, we can also avoid checking this. But if we want to check this, I feel we need an explicit list of supported language codes and then check if the exploration's language code is present in that list.

tjiang11 replied:
Okay, yes I think that would be fine (having our own list of supported languages).

bansalnitish replied:
So, just to confirm once:

- We have to store a list of webkit supported languages
- If the code for exploration language is found in the list, we provide the option for speech recognition otherwise don't.

bansalnitish replied:
The list of the languages would be quite long though.

tjiang11 replied:
I think that's okay, you can see in constants.js there is quite a long list of languages.

bansalnitish replied:
Also, should I add a list of only language codes or codes with their description since for checking only the codes would be needed?

tjiang11 replied:
Whichever is necessary to perform the check, so I think just the language codes is fine

tjiang11 reviewed (commented):
Hi @bansalnitish , one more question. Also, I think there's one earlier comment that is still unresolved about indentation.

bansalnitish commented:
Hi @tjiang11, I have added the language codes for checking if a language is supported in speech recognition. Also, I have added the description for making the list more readable, PTAL!

tjiang11 commented:
Hi @bansalnitish , it seems like it's breaking--it looks like you're trying to use ExplorationLanguageCodeService but never defined it.

bansalnitish commented:
Hi @tjiang11, I have added ExplorationLanguageCodeService in the controller as [here](https://github.com/oppia/oppia/pull/4740/files#diff-dd4f17b179e3fa64d8bd29af3b53b406R37) and it works well locally. Can you please guide me here, I didn't get you :sweat_smile:  Thanks!

tjiang11 commented on core/templates/dev/head/components/forms/schema_editors/SchemaBasedUnicodeEditorDirective.js:
```diff
@@ -121,6 +123,65 @@ oppia.directive('schemaBasedUnicodeEditor', [
+            ExplorationLanguageCodeService.displayed);
+
+          var isSpeechRecognitionLanguagesupported = (
+            constants.SUPPORTED_SPEECH_RECOGNITION_LANGUAGES.filter(
```
Correct me if I'm wrong but doesn't filter return another array?

bansalnitish replied:
Yes, it does but I was checking if the returned array is empty or not. I have now shifted the check to this same line for avoiding confusion!

tjiang11 commented:
Oops, had some hiccups on my end with my environment. Anyways, it looks like the microphone looks like it got a bit off-centered, can you take a look?

![image](https://user-images.githubusercontent.com/11807091/37313274-dd2d0cdc-2624-11e8-9430-7acc25d5f1cb.png)

bansalnitish commented:
Hi @tjiang11, I have fixed the position of the mic!

![image](https://user-images.githubusercontent.com/22434689/37345042-02024e82-26f2-11e8-882c-71bbb811046d.png)

tjiang11 commented:
Hi @bansalnitish , I just realized--you're using ExplorationLanguageCodeService, but that's a service for the editor, not the player.

bansalnitish commented:
Hi @tjiang11, Sorry for the delay on this. I was trying to use ExplorationPlayerService for this but it does not seem to work here. Can you please help me out here. Thanks!

tjiang11 commented:
Hi @bansalnitish , sorry for the late response--I was out of town. It looks like Exploration objects have a 'languageCode' property (refer to ExplorationObjectFactory.js). Perhaps you can use that?

bansalnitish commented:
Hi @tjiang11, I have made some changes but I am stuck with the failing tests since many days and I am unable to resolve the issue. Can you please guide me!

brianrodri commented on core/templates/dev/head/components/forms/schema_editors/SchemaBasedUnicodeEditorDirective.js:
```diff
@@ -121,6 +123,67 @@ oppia.directive('schemaBasedUnicodeEditor', [
+          var explorationLanguageCode = (
+            explorationData.$$state.value.language_code);
+
+          var isSpeechRecognitionLanguagesupported = (
```
Please capitalize the s in supported: `isSpeechRecognitionLanguageSupported`

brianrodri commented on core/templates/dev/head/components/forms/schema_editors/SchemaBasedUnicodeEditorDirective.js:
```diff
@@ -121,6 +123,67 @@ oppia.directive('schemaBasedUnicodeEditor', [
+            constants.SUPPORTED_SPEECH_RECOGNITION_LANGUAGES.filter(
+              function(languageCodeDict) {
+                return languageCodeDict.code === explorationLanguageCode;
+              }).length !== 0);
```
Use [`some`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) instead?
```
constants.SUPPORTED_SPEECH_RECOGNITION_LANGUAGES.some(
    function(languageCodeDict) {
      return languageCodeDict.code === explorationLanguageCode;
    })
```

brianrodri commented on core/templates/dev/head/components/forms/schema_editors/schema_based_unicode_editor_directive.html:
```diff
@@ -1,13 +1,15 @@
-            ng-focus="onInputFocus()">
-  </textarea>
+  <span ng-if="!getCodingMode() && getRows()">
+    <textarea ng-if="!getCodingMode() && getRows()"
```
This `ng-if` is no longer necessary since the parent `span` tag uses it.

bansalnitish commented:
Hi @brianrodri, sorry for the delay on this. I have made the requested changes.

bansalnitish commented:
Hi, @seanlip, @tjiang11 I need some help from you here. This PR is opened since many days and I am stuck with the failing build. Can anyone of you please have a look and guide me here?

seanlip commented:
Did you read the Travis logs? You can see them by clicking "Details" in the summary above.

The failures all seem to be due to some microphone.png issue. Take a look elsewhere in the codebase to see how images are included. There are certain patterns that need to be followed here for the image loading to work correctly in both dev and prod.

bansalnitish commented:
Yes I do read the travis logs for the build failure but here I didn't get why the image is not being fetched. I'll take a look at the codebase once more to see if there are any specific patterns (which I haven't followed) for the image.
Thanks @seanlip

seanlip commented:
Ah ok, sounds good, thanks.

Btw, for future reference -- when you're stuck, it's a good idea to talk about where exactly you're stuck, and refer to specific artifacts / files / links, to narrow the scope of the "issue location" as much as possible. Otherwise, the person responding on the other end hasn't really got any idea what the difficulty is, and might give you info you already know. (No worries about this case; just mentioning it as a note for the future!)

bansalnitish commented:
Yeah! I'll take care in future.
Thanks!

bansalnitish commented:
Hi @seanlip, I have fixed the issue with the images.

Another issue which I am facing is that the embedded explorations with text input interaction are not being played. This is due to [unknown provider](https://docs.angularjs.org/error/$injector/unpr) error for ExplorationLanguageCodeService which I have used to obtain the language code for the current exploration. I tried to find a fix but I am stuck with this. Can you please guide me here. Thanks!

seanlip commented:
Probably the necessary service isn't being imported in the relevant HTML file (in the JS script tags). That's usually the cause of these issues.

bansalnitish commented:
Hi @seanlip, I fixed the issue locally by adding the necessary script tags for ExplorationLanguageCodeService but I am not using the right service for finding the language code here. Can you please guide me on how can I obtain the language code for the current exploration.

seanlip commented:
Look into explorationPropertyService and its subclassea.

bansalnitish commented:
I have used ExplorationLanguageCodeService which is a subclass of ExplorationPropertyService but @tjiang11 suggested to try a different approach as in [here](https://github.com/oppia/oppia/pull/4740#issuecomment-375841764). I am not sure about which service should I use other than this.

seanlip commented:
The ExplorationLanguageCodeService is a service that wraps the current exploration's language code property. If you need to get the current language code for the exploration, and you don't have the exploration object readily accessible, then using that service is the way to go.

bansalnitish commented:
Hi @seanlip, I get an error as here: https://travis-ci.org/oppia/oppia/jobs/368782042#L8664. I am unable to understand the reason behind this. Can you please me guide me here, Thanks!

seanlip commented:
I'm afraid I have no idea. But if the corresponding tests run fine locally, feel free to go ahead and restart the failing test. See https://github.com/oppia/oppia/wiki/If-your-build-fails

bansalnitish commented:
I have already restarted the build (multiple times before) but the error persists. I am looking into this and I will try to find out the issue.

bansalnitish commented:
Hi @seanlip @apb7, I have fixed the issues with all other tests. All of them pass except the frontend tests. I ran the test locally and it works well. I am unable to figure out the issue here: https://travis-ci.org/oppia/oppia/jobs/369675427#L8788. 
Can you please help me with this, Thanks!

apb7 commented:
Looking into this. Thanks, @bansalnitish!

seanlip commented:
I can't look at this now, but have you tried searching for the error message @bansalnitish?

apb7 commented:
Hi @bansalnitish! I found a similar question on StackOverflow [here](https://github.com/oppia/oppia/pull/4740/files). Does this help?
Thanks!

anmolshkl commented:
Hi @bansalnitish, were you able to resolve the issue or are you still stuck on this?

bansalnitish commented:
Hi @anmolshkl, I was busy with my exams. I am looking into this now and will update you if I am unable to fix this, Thanks!

bansalnitish commented:
Hi @seanlip, Yes I have searched for the error but the results I got are not linked to the error. I am looking into this once again and will update you if I am unable to fix this.

Hi @apb7, thanks for the [reference](https://stackoverflow.com/questions/43686810/module-ngmaterial-is-not-available-you-either-misspelled-the-module-name-or-f), I am trying to figure out the reason behind this error so that I can fix it.

bansalnitish commented:
Hi @anmolshkl, I ran the frontend tests locally after reverting all changes and also on develop branch. I get same error there as well. I am looking into this. Can you please run the frontend tests locally and update me with the results, Thanks!

kevinlee12 commented:
Hi @bansalnitish, did you get around to finding out why it was erroring? thanks!

1995YogeshSharma commented:
Hi @bansalnitish , any updates on this PR?

bansalnitish commented:
HI @kevinlee12 @1995YogeshSharma, sorry for the delay in reply. I was unable to find the reason behind the errors. Also, presently I am working on TextAngular migration, so was unable to look into this. Once that is done, I will once again try to find the reason behind the error.
Thanks!

bansalnitish commented:
Yes I'll work on this after the completion of my phase 2.
